File: src/serialization/json_serializer.cpp
#include "json_serializer.hpp"
#include <json/json.h>
#include <iostream>

namespace trame {

json_serializer::json_serializer() : writer(new Json::FastWriter) {

}

json_serializer::json_serializer(const json_serializer&) {

}

json_serializer::~json_serializer() {
    delete writer;
}

std::vector<unsigned char> json_serializer::serialize(skeleton s) {
    Json::Value node;

    node["timestamp"] = s.timestamp;
    node["id"] = s.id;
    node["root"] = object_from_joint(s.root);

    std::string serialized = writer->write(node);
    std::vector<unsigned char> bytes(serialized.begin(), serialized.end());

    return bytes;
}

output_type json_serializer::get_type() {
    return output_type::JSON;
}

Json::Value json_serializer::object_from_joint(joint j) {
    Json::Value joint_obj;
    joint_obj["type"] = (int)j.type;
    joint_obj["normal"] = array_from_vector(j.normal);
    joint_obj["point"] = array_from_vector(j.point);
    Json::Value children(Json::arrayValue);

    for (joint child : j.children) {
        children.append(object_from_joint(child));
    }

    joint_obj["children"] = children;
    return joint_obj;
}

Json::Value json_serializer::array_from_vector(Eigen::Vector3d v) {
    if(v.norm() > 0) {
        Json::Value vector_arr;
        vector_arr.append(v(0));
        vector_arr.append(v(1));
        vector_arr.append(v(2));

        return vector_arr;
    } else {
        return Json::Value(Json::nullValue);
    }
}

}

File: csharp/TrameSerialization/Serializer/JSONSerializer.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Json;
using System.Linq;
using System.Text;
using AForge.Math;
using Trame;


namespace TrameSerialization.Serializer
{
    class JSONSerializer : ISerializer
    {
        public OutputType OutputType
        {
            get { return OutputType.JSON; }
        }

        public Stream Serialize(Trame.ISkeleton<Vector4, Vector3> s)
        {
            var stream = new MemoryStream();
            var writer = new StreamWriter(stream);
            var jsonSkeleton = new JsonObject();

            jsonSkeleton["id"] = s.ID;
            jsonSkeleton["timestamp"] = s.Timestamp;
            jsonSkeleton["valid"] = s.Valid;
            jsonSkeleton["root"] = JointToObject(s.Root);

            writer.Write(jsonSkeleton.ToString());
            writer.Flush();
            stream.Position = 0;
            return stream;;
        }

        public ISkeleton<Vector4, Vector3> Deserialize(Stream stream)
        {
            throw new NotImplementedException();
        }

        private static JsonValue JointToObject(IJoint<Vector4, Vector3> joint)
        {
            if (joint == null)
            {
                return null;
            }
            var jsonJoint = new JsonObject();
            jsonJoint["type"] = (int) (Convert.ChangeType(joint.JointType, TypeCode.Int32));
            jsonJoint["orientation"] = Vector4ToArray(joint.Orientation);
            jsonJoint["point"] = Vector3ToArray(joint.Point);
            jsonJoint["valid"] = joint.Valid;
            var children = new JsonArray();

            foreach (var child in joint.GetChildren())
            {
                children.Add(JointToObject(child));
            }

            jsonJoint["children"] = children;

            return jsonJoint;
        }

        private static JsonArray Vector3ToArray(Vector3 vector)
        {
            if (vector.Norm > 0)
            {
                var arr = new JsonArray {vector.X, vector.Y, vector.Z};
                return arr;
            }
            else
            {
                return null;
            }
        }

        private static JsonArray Vector4ToArray(Vector4 vector)
        {
            if (vector.Norm > 0)
            {
                var arr = new JsonArray {vector.X, vector.Y, vector.Z, vector.W};
                return arr;
            }
            else
            {
                return null;
            }
        }
    }
}
